Utilities/rANS/include/rANS/internal/Encoder.h

```diff
@@ -33,154 +33,119 @@ namespace rans
-  State_T x;
-  std::tie(x, streamPos) = renorm(mState, iter, freq, scale_bits);
-
-  // x = C(s,x)
-  mState = ((x / freq) << scale_bits) + (x % freq) + start;
-  return streamPos;
-};
-
-template <typename State_T, typename Stream_T>
-template <typename Stream_IT, std::enable_if_t<isCompatibleIter_v<Stream_T, Stream_IT>, bool>>
-Stream_IT Encoder<State_T, Stream_T>::flush(Stream_IT iter)
-{
-
-  Stream_IT streamPos = iter;
-
-  if constexpr (needs64Bit<State_T>()) {
-    ++streamPos;
-    *streamPos = static_cast<Stream_T>(mState >> 32);
-    ++streamPos;
-    *streamPos = static_cast<Stream_T>(mState >> 0);
-    assert(std::distance(iter, streamPos) == 2);
+  stream_IT streamPosition = outputIter;
+  if constexpr (needs64Bit<state_T>()) {
+    ++streamPosition;
+    *streamPosition = static_cast<stream_T>(mState >> 32);
+    ++streamPosition;
+    *streamPosition = static_cast<stream_T>(mState >> 0);
   } else {
-    ++streamPos;
-    *streamPos = static_cast<Stream_T>(mState >> 24);
-    ++streamPos;
-    *streamPos = static_cast<Stream_T>(mState >> 16);
-    ++streamPos;
-    *streamPos = static_cast<Stream_T>(mState >> 8);
-    ++streamPos;
-    *streamPos = static_cast<Stream_T>(mState >> 0);
-    assert(std::distance(iter, streamPos) == 4);
+    ++streamPosition;
+    *streamPosition = static_cast<stream_T>(mState >> 24);
+    ++streamPosition;
+    *streamPosition = static_cast<stream_T>(mState >> 16);
+    ++streamPosition;
+    *streamPosition = static_cast<stream_T>(mState >> 8);
+    ++streamPosition;
+    *streamPosition = static_cast<stream_T>(mState >> 0);
   }
-
   mState = 0;
-  return streamPos;
+  return streamPosition;
 };
 
-template <typename State_T, typename Stream_T>
-template <typename Stream_IT, std::enable_if_t<isCompatibleIter_v<Stream_T, Stream_IT>, bool>>
-Stream_IT Encoder<State_T, Stream_T>::putSymbol(Stream_IT iter, const EncoderSymbol<State_T>& sym, uint32_t scale_bits)
+template <typename state_T, typename stream_T>
+template <typename stream_IT>
+stream_IT Encoder<state_T, stream_T>::putSymbol(stream_IT outputIter, const EncoderSymbol<state_T>& symbol)
 {
 
-  assert(sym.freq != 0); // can't encode symbol with freq=0
+  assert(symbol.getFrequency() != 0); // can't encode symbol with freq=0
 
   // renormalize
-  Stream_IT streamPos;
-  State_T x;
-  std::tie(x, streamPos) = renorm(mState, iter, sym.freq, scale_bits);
+  const auto [newState, streamPosition] = renorm(mState, outputIter, symbol.getFrequency());
 
   // x = C(s,x)
-  State_T q = 0;
+  state_T quotient = 0;
 
-  if constexpr (needs64Bit<State_T>()) {
+  if constexpr (needs64Bit<state_T>()) {
     // This code needs support for 64-bit long multiplies with 128-bit result
     // (or more precisely, the top 64 bits of a 128-bit result).
-    q = static_cast<State_T>((static_cast<uint128>(x) * sym.rcp_freq) >> 64);
+    quotient = static_cast<state_T>((static_cast<uint128_t>(newState) * symbol.getReciprocalFrequency()) >> 64);
```

Comment:
For reference, I just did some experiments with how GCC compiles this code.
Actually already starting from -O1 it optimizes the 128 bit away and does only a 64 bit multiplication and takes the higher bits of the result. Only thing that is important is that one always casts a 64 bit value to 128 bit directly in the multiplication, such that the compiler knows the upper 64 bit are empty. I.e. if you store the `newState` as int128 variable, that blows up the assembler code.
Then, there is also the umul / umulh / etc. intrinsic, but apparently there is consensus among the compilers on that, so I think the way you do it here is already as good as it gets. And nice to see that it does not yield any overhead. Perhaps you can add a comment somewhere, that despite using 128 bit, you are actually only doing a 64x64 bit multiplication, taking the upper 64 bit of the result, and the compiler optimizes everything away so it just uses amd64 integer instructions.